### PR TITLE
VKT: Optimizations

### PR DESCRIPTION
**Pull request by jsign**
This PR contains more optimizations apart from the ones in:
- https://github.com/gballet/go-verkle/pull/314
- https://github.com/crate-crypto/go-ipa/pull/31

The `go.mod` in this PR uses those branches; we should update the official commits whenever those PRs are merged.

Apart from that, we're doing other optimizations that will be explained in PR comments.

The results are significant comparing `beverly-hills` vs this PR:
```
name                               old time/op    new time/op    delta
TriesRandom/VKT/1000_accounts-16     56.8ms ± 3%    22.9ms ± 3%  -59.73%  (p=0.001 n=10+5)
TriesRandom/VKT/5000_accounts-16      264ms ± 2%      76ms ± 2%  -71.34%  (p=0.001 n=10+5)
TriesRandom/VKT/10000_accounts-16     535ms ± 2%     146ms ± 5%  -72.78%  (p=0.001 n=10+5)

name                               old alloc/op   new alloc/op   delta
TriesRandom/VKT/1000_accounts-16     9.56MB ± 0%    9.76MB ± 0%   +2.13%  (p=0.001 n=9+5)
TriesRandom/VKT/5000_accounts-16     43.5MB ± 0%    43.9MB ± 0%   +0.94%  (p=0.002 n=8+5)
TriesRandom/VKT/10000_accounts-16    87.5MB ± 0%    88.2MB ± 0%   +0.83%  (p=0.001 n=10+5)

name                               old allocs/op  new allocs/op  delta
TriesRandom/VKT/1000_accounts-16      32.8k ± 0%     35.5k ± 0%   +8.08%  (p=0.001 n=9+5)
TriesRandom/VKT/5000_accounts-16       151k ± 0%      144k ± 0%   -4.88%  (p=0.001 n=10+5)
TriesRandom/VKT/10000_accounts-16      303k ± 0%      285k ± 0%   -5.97%  (p=0.001 n=10+5)
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -106,3 +106,7 @@ require (
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+replace github.com/gballet/go-verkle => github.com/gballet/go-verkle v0.0.0-20221213233340-474c89b28152
+
+replace github.com/crate-crypto/go-ipa => github.com/jsign/go-ipa v0.0.0-20221213212217-fcc1cb4ffdd5
```
Comment (jsign) [inline go.mod:112]: To be changed again when the dependency PRs get merged.

File: tests/tries_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package tests
```
Comment (jsign) [inline tests/tries_test.go:1]: Here we have the benchmarks from https://github.com/gballet/go-ethereum/pull/146.
I only removed the statelessness one as we discussed, so we avoid test flags.

File: trie/utils/verkle.go [hunk 3]
```diff
@@ -82,9 +82,11 @@ func GetTreeKey(address []byte, treeIndex *uint256.Int, subIndex byte) []byte {
...
 	// - poly[4]'s byte representation is the same as the *low* 16 bytes (trieIndexBytes[:16]) of
 	//   the 32-byte aligned big-endian representation (BE({00,00,...}).
-	trieIndexBytes := treeIndex.Bytes32()
-	verkle.FromBytes(&poly[3], trieIndexBytes[16:])
-	verkle.FromBytes(&poly[4], trieIndexBytes[:16])
+	if !treeIndex.IsZero() {
+		trieIndexBytes := treeIndex.Bytes32()
+		verkle.FromBytes(&poly[3], trieIndexBytes[16:])
+		verkle.FromBytes(&poly[4], trieIndexBytes[:16])
+	}
```
Comment (jsign) [inline trie/utils/verkle.go:89]: First optimization is avoiding work if `treeIndex` is zero, which makes sense and is a pretty normal case.

File: trie/verkle.go [hunk 4]
```diff
@@ -119,7 +119,7 @@ func (t *VerkleTrie) TryUpdateAccount(key []byte, acc *types.StateAccount) error
 	var (
 		err            error
 		nonce, balance [32]byte
-		values         = make([][]byte, verkle.NodeWidth)
+		values         = make([][]byte, 4)
```
Comment (jsign) [inline trie/verkle.go:122]: This change was a big part of our memory savings.

Pre-allocating 256 slice of slices means that we are reserving (24 bytes * 256)=6KiB of memory that will have to be garbage collected later.

If you look at the code below, in this `TryUpdateAccount` we only care about the first four slots.
This is exploiting a fact that `InsertAtStem(..)` doesn't require fixed sized slices, since if the value slice is smaller the `go-verkle` code still works correctly (since it iterates the slice of leaf values to do hashes).

If this optimization feels too aggressive, we can think about reverting it but I remember it had a significant impact in memory consumption that increased pressure in GC. We can discuss it further.

File: trie/verkle.go [hunk 5]
```diff
@@ -156,10 +156,9 @@ func (t *VerkleTrie) TryUpdateAccount(key []byte, acc *types.StateAccount) error
 }
 
 func (trie *VerkleTrie) TryUpdateStem(key []byte, values [][]byte) {
-	resolver :=
-		func(h []byte) ([]byte, error) {
-			return trie.db.diskdb.Get(h)
-		}
+	resolver := func(h []byte) ([]byte, error) {
+		return trie.db.diskdb.Get(h)
+	}
```
Comment (jsign) [inline trie/verkle.go:161]: Lint.

File: trie/verkle.go [hunk 6]
```diff
@@ -239,9 +238,23 @@ func nodeToDBKey(n verkle.VerkleNode) []byte {
...
+		value, err := n.Serialize()
+		if err != nil {
+			panic(err)
+		}
+		dbKey := nodeToDBKey(n)
+		flush <- vnflush{
+			n:     n,
+			value: value,
+			dbKey: dbKey,
+		}
```
Comment (jsign) [inline trie/verkle.go:257]: Looking further at where the wall-clock time was going, I realized that after `Commit()`ing the underlying `go-verkle` trie, a significant amount of time was spent flushing the result.

What I did in https://github.com/gballet/go-verkle/pull/314, is making `Flush()` for the root node to do the work in parallel. This means that `resolver` will be called by multiple goroutines balancing more CPU work in all available cores.

To squeeze things more, the `Serialize()` part of the process was now moved from L258 to the `resolver`. This means that the `Serialize()` work will also be done in the goroutines that are flushing the result, and not in a single core.

In a nutshell, since the resolver execution is exploiting all cores, we want to do as much work as possible there, so the main goroutine ranging L269 is receiving plain results to be stored in the `diskdb`. This main goroutine isn't doing heavy CPU work, just receiving exactly what needs to be stored. 

This is important, since if the `range` in L269 is slow, it will slow down everything. Note how I also made a buffered channel in L246. We need to avoid as much as possible to be blocking goroutines, so having extra breathing room also improved performance.

Comment (gballet) [reply trie/verkle.go:257]: So if #295 gets merged, `Serialize` will become a tad more expensive - at the cost of a cheaper deserialization (hence not producing a tree that is crash-prone because of potentially invalid commitments, as in your current approach). Your change, I believe, will mitigate this problem and make it cheaper to follow this approach. That's very nice.

**Comment by jsign** [comment]
Should be interesting if we consider https://github.com/gballet/go-verkle/pull/314, but for now I'll close this PR.
We can recover it if we go that route.
